packages/fuels-core/src/abi_decoder.rs

```diff
@@ -1,252 +1,287 @@
+    }
+
+    fn decode_u8(data: &[u8]) -> Result<DecodeResult, CodecError> {
+        let slice = peek_word(data)?;
+
+        let result = DecodeResult {
+            token: Token::U8(u8::from_be_bytes(slice[7..8].try_into().unwrap())),
+            bytes_read: 8,
+        };
+
+        Ok(result)
+    }
+
+    fn decode_unit() -> Result<DecodeResult, CodecError> {
+        let result = DecodeResult {
+            token: Token::Unit,
+            bytes_read: 0,
+        };
+        Ok(result)
+    }
+
+    /// Will decode an Enum according to TODO: PUT_FUEL_SPECS_ENUM_ENCODING_LINK_HERE
+    ///
+    /// # Arguments
+    ///
+    /// * `data`: slice of encoded data on whose beginning we're expecting an encoded enum
+    /// * `variants`: all types that this particular enum type could hold
+    fn decode_enum(data: &[u8], variants: &EnumVariants) -> Result<DecodeResult, CodecError> {
+        let discriminant = Self::decode_discriminant(data)?;
+        let bytes_read = ENUM_DISCRIMINANT_WORD_WIDTH * WORD_SIZE;
+
+        let variant = Self::select_variant(variants, discriminant as usize)?;
+        let res = Self::decode_param(variant, &data[bytes_read..])?;
+
+        let selector = Box::new((discriminant as u8, res.token, variants.clone()));
+        let token = Token::Enum(selector);
+
+        let enum_width = encoding_width_to_fit_any(variants) + ENUM_DISCRIMINANT_WORD_WIDTH;
+        Ok(DecodeResult {
+            token,
+            bytes_read: enum_width * WORD_SIZE,
+        })
+    }
+
+    fn decode_discriminant(data: &[u8]) -> Result<u32, CodecError> {
+        let discriminant = peek_word(data)?;
+        Ok(u32::from_be_bytes(discriminant[4..8].try_into().unwrap()))
+    }
+
+    /// Returns a variant from `variants` pointed to by `discriminant`.
+    /// Will fail if `discriminant` is out of bounds.
+    fn select_variant(
+        variants: &EnumVariants,
+        discriminant: usize,
+    ) -> Result<&ParamType, CodecError> {
+        variants.param_types().get(discriminant).ok_or_else(|| {
+            let msg = format!(
+                concat!(
+                    "Error while decoding an enum. The discriminant '{}' doesn't ",
+                    "point to any of the following variants: {:?}"
+                ),
+                discriminant, variants
+            );
+            CodecError::InvalidData(msg)
+        })
     }
 }
 
-fn peek(data: &[u8], offset: usize, len: usize) -> Result<&[u8], CodecError> {
-    if offset + len > data.len() {
+fn peek(data: &[u8], len: usize) -> Result<&[u8], CodecError> {
+    if len > data.len() {
         Err(CodecError::InvalidData(
             "requested data out of bounds".into(),
         ))
     } else {
-        Ok(&data[offset..(offset + len)])
+        Ok(&data[..len])
     }
 }
 
-fn peek_word(data: &[u8], offset: usize) -> Result<ByteArray, CodecError> {
-    peek(data, offset, WORD_SIZE as usize).map(|x| {
+fn peek_word(data: &[u8]) -> Result<ByteArray, CodecError> {
```

Comment:
Even if it seems to work since the tests pass, I don't understand why you can simply not care about any offset anymone?

Reply:
A cool trick.

Slices are nothing more than a ptr and a length

So I removed the '+offset' everywhere and just opted to move the slice further right along the data as we continue to decode.

The only thing that is now important is that each decoding step says how many bytes they've already consumed. 

Using that info the root function from which we started (decode) will construct an appropriate slice for the next decoding iteration.

This way we're always reading from the beginning and the offset is encoded in the type :D

Reply:
Smart! This should be documented somewhere.